Sources/Turf/GeoJSON.swift

```diff
@@ -0,0 +1,358 @@
+                let feature = try mappedContainer.decode(type.self)
+                features.append(feature)
+            }
+        }
+        
+        self.features = features
+        
+        try super.init(from: decoder)
+    }
+    
+    public override func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(self.properties, forKey: .properties)
+        try container.encode(self.features, forKey: .features)
+        
+        try super.encode(to: encoder)
+    }
+}
+
+public enum GeometryType: String {
+    case Point
+    case LineString
+    case Polygon
+    case MultiPoint
+    case MultiLineString
+    case MultiPolygon
+    
+    static let allValues: [GeometryType] = [.Point, .LineString, .Polygon, .MultiPoint, .MultiLineString, .MultiPolygon]
+    
+    var featureType: Feature.Type {
+        switch self {
+        case .LineString:
+            return LineStringFeature.self
+        case .Polygon:
+            return PolygonFeature.self
+        case .Point:
+            return PointFeature.self
+        case .MultiPoint:
+            return MultiPointFeature.self
+        case .MultiLineString:
+            return MultiLineStringFeature.self
+        case .MultiPolygon:
+            return MultiPolygonFeature.self
+        }
+    }
+}
+
+public enum GeoJSONType: String {
+    case Feature
+    case FeatureCollection
+    static let allValues: [GeoJSONType] = [.Feature, .FeatureCollection]
+}
+
+public enum GeoJSONError: Error {
+    case unknownType
+    case noTypeFound
+}
+
+
+
+public class GeoJSON: Codable {
+    
+    var value: Codable?
+    
+    public required init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let feature = try container.decode(Feature.self)
+        
+        if feature.type == .Feature {
+            guard let type = GeometryType(rawValue: feature.simplifiedGeometry!.type) else {
+                throw GeoJSONError.unknownType
+            }
+            self.value = try container.decode(type.featureType)
+        } else if feature.type == .FeatureCollection {
+            self.value = try container.decode(FeatureCollection.self)
+        } else {
+            throw GeoJSONError.noTypeFound
+        }
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        
+        if let value = value as? FeatureCollection {
+            try container.encode(value)
+        } else if let value = value as? Feature {
+            try container.encode(value)
+        } else {
+            throw GeoJSONError.unknownType
+        }
+    }
+    
+    public static func parse(data: Data) throws -> GeoJSON {
+        return try JSONDecoder().decode(GeoJSON.self, from: data)
+    }
+    
+    public static func parse<T: BaseFeature>(data: Data, as: T.Type) throws -> T {
```

Comment:
What if we change this around so that you call e.g. `FeatureCollection(geoJSONData:)` if you expect it to be a feature collection?

Reply:
Could help with discoverability but it would bloat the code a bit because it would mean one implementation per feature/geometry. I figured a similar signature to JSONDecoder’s decode would make more sense so I renamed this to `GeoJSON.parse(_:from:)`